docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,473 @@
+
+There's a straw-man proposal for this at the end of the document.
+
+## A command-line (and possibly Android) utility.
+
+We should build a utility (or 2) which can stand in for the high-level
+component, for testing and demonstration purposes.
+
+This is something like places-utils.rs and the little utility Grisha has
+been using. This utility should act like a real client (ie, it should have
+an FxA device record, etc) and it should use the low-level component in
+exactly the same we we expect real products to use it.
+
+Because it is just a consumer of the low-level component, it will force us to
+confront some key issues, such as how to get references to engines stored in
+multiple crates, how to present a unified "state" for things like auth errors,
+etc.
+
+## The "clients" engine
+
+The initial work for the clients engine can probably be done without too
+much regard for how things are tied together - for example, much work could
+be done without caring how we get a reference to engines across crates.
+
+## State work
+
+Implementing things needed to we can expose the correct state to the high-level
+manager for things like auth errors, backoff semantics, etc
+
+## Tie it together and other misc things.
+
+There will be lots of loose ends to clean up - things like telemetry, etc.
+
+# Followup with non-rust engines.
+
+We have identified that iOS will, at least in the short term, want the
+sync manager to be implemented in Swift. This will be responsible for
+syncing both the Swift and Rust implemented engines.
+
+At some point in the future, Desktop may do the same - we will have both
+Rust and JS implemented engines which need to be coordinated. We ignore this
+requirement for now.
+
+This approach still has a fairly easy time coordinating with the Rust
+implemented engines - the FFI will need to expose the relevant sync
+entry-points to be called by Swift, but the Swift code can hard-code the
+Rust engines it has and make explicit calls to these entry-points.
+
+This Swift code will need to create the structures identified below, but this
+shouldn't be too much of a burden as it already has the information necessary
+to do so (ie, it already has info/collections etc)
+
+TODO: dig into the Swift code and make sure this is sane.
+
+# Details
+
+While we use rust struct definitions here, it's important to keep in mind that
+as mentioned above, we'll need to support the manager being written in
+something other than rust, and to support engines written in other than rust.
+
+The structures below are a straw-man, but hopefully capture all the information
+that needs to be passed around.
+
+```rust
+
+// We want to define a list of "engine IDs" - ie, canonical strings which
+// refer to what the user perceives as an "enigine" - but as above, these
+// *do not* correspond 1:1 with either "stores" or "collections" (eg, "history"
+// refers to 2 stores, and in a future world, might involve 3 collections).
+enum Engine {
```

Comment:
We've inherited this name from the existing sync system, but this might also be a good opportunity to bikeshed a better one. I can't come up with a good rationale for why this is called an "engine", particularly in a world where most of the driving of a sync is done by the store.